src/support/z_ssl_certs.erl

```diff
@@ -0,0 +1,329 @@
+            {error, {cannot_read_pemfile, Filename, Error}}
+    end.
+
+-spec generate_self_signed(string()|binary(), list()) -> {ok, list()} | {error, term()}.
+generate_self_signed(NormalizedHostname, Opts) ->
+    PemFile = proplists:get_value(keyfile, Opts),
+    lager:info("Generating self-signed ssl keys for ~s in ~s", [NormalizedHostname, PemFile]),
+    case filelib:ensure_dir(PemFile) of
+        ok ->
+            KeyFile = filename:rootname(PemFile) ++ ".key",
+            CertFile = proplists:get_value(certfile, Opts),
+            Command = "openssl req -x509 -nodes"
+                    ++ " -days 3650"
+                    ++ " -sha256"
+                    ++ " -subj '/CN="++z_convert:to_list(NormalizedHostname)++"'"
+                    ++ " -newkey rsa:"++?BITS++" "
+                    ++ " -keyout "++z_utils:os_filename(KeyFile)
+                    ++ " -out "++z_utils:os_filename(CertFile),
+            lager:debug("SSL: ~p", [Command]),
+            Result = os:cmd(Command),
+            lager:debug("SSL: ~p", [Result]),
+            case file:read_file(KeyFile) of
+                {ok, <<"-----BEGIN PRIVATE KEY", _/binary>>} ->
+                    os:cmd("openssl rsa -in "++KeyFile++" -out "++PemFile),
+                    {ok, Opts};
+                {ok, <<"-----BEGIN RSA PRIVATE KEY", _/binary>>} ->
+                    file:rename(KeyFile, PemFile),
+                    {ok, Opts};
+                _Error ->
+                    lager:error("Failed generating self-signed ssl keys for ~s in ~s (output was ~s)",
+                                [NormalizedHostname, PemFile, Result]),
+                    {error, openssl}
+            end;
+        {error, _} = Error ->
+            {error, {ensure_dir, Error, PemFile}}
+    end.
+
+%% @doc Return the dh key to be used. Needed for better forward secrecy with the DH key exchange
+-spec dh_options() -> [ssl:ssl_option()].
+dh_options() ->
+    [{dhfile, dhfile()}].
+
+is_dhfile(Filename) ->
+    case file:read_file(Filename) of
+        {ok, <<"-----BEGIN DH PARAMETERS", _/binary>>} -> true;
+        _ -> false
+    end.
+
+ensure_dhfile() ->
+    ensure_dhfile(dhfile()).
+
+dhfile() ->
+    case z_config:get(ssl_dhfile) of
+        undefined -> filename:join([z_utils:lib_dir(priv), "ssl", "dh"++?DHBITS++".pem"]);
+        Filename -> Filename
+    end.
+
+ensure_dhfile(Filename) ->
+    case filelib:is_file(Filename) of
+        true ->
+            ok;
+        false ->
+            ok = filelib:ensure_dir(Filename),
+            % Maybe use the -dsaparam, maybe we shouldn't https://www.openssl.org/news/secadv/20160128.txt
+            % Though without this the generation will take a long time indeed...
+            lager:info("Generating ~s bits DH key, this will take a long time (saving to ~p)",
+                       [?DHBITS, Filename]),
+            Command = "openssl dhparam -out " ++ z_utils:os_filename(Filename) ++ " " ++ ?DHBITS,
```

Comment:
This is a new dependency on openssl, right? We need to add it to the docs and Docker images.